packages/cursorless-engine/src/core/updateSelections/updateSelections.ts

```diff
@@ -14,33 +14,114 @@ import {
 import { performDocumentEdits } from "../../util/performDocumentEdits";
 import { RangeUpdater } from "./RangeUpdater";
 
+type SelectionsOrRanges = readonly Selection[] | readonly Range[];
+
 interface SelectionsWithBehavior {
-  selections: readonly Selection[];
-  rangeBehavior?: RangeExpansionBehavior;
+  selections: SelectionsOrRanges;
+  behavior: RangeExpansionBehavior;
 }
 
-/**
- * Given a selection, this function creates a `SelectionInfo` object that can
- * be passed in to any of the commands that update selections.
- *
- * @param document The document containing the selection
- * @param selection The selection
- * @param rangeBehavior How selection should behave with respect to insertions on either end
- * @returns An object that can be used for selection tracking
- */
-export function getSelectionInfo(
+interface BaseProps<K extends string> {
+  rangeUpdater: RangeUpdater;
+  editor: EditableTextEditor;
+  preserveEditorSelections?: boolean;
+  selections: Record<K, SelectionsOrRanges | SelectionsWithBehavior>;
+}
+
+interface EditsProps<K extends string> extends BaseProps<K> {
+  edits: Edit[];
+}
+
+interface CallbackProps<K extends string> extends BaseProps<K> {
+  callback: () => Promise<void>;
+}
+
+type UpdaterProps<K extends string> = EditsProps<K> | CallbackProps<K>;
+
+export async function performEditsAndUpdateSelections<K extends string>({
+  rangeUpdater,
+  editor,
+  selections,
+  preserveEditorSelections,
+  ...rest
+}: UpdaterProps<K>): Promise<Record<K, Selection[]>> {
+  const keys = Object.keys(selections) as K[];
```

Comment:
I think you can use our `unsafeKeys` utility here

Reply:
Good call